### PR TITLE
Move merge selected features logic to QgsVectorLayerEditUtils

### PR DESCRIPTION
**Pull request by domi4484**
Currently the logic of the merge selected features tool is implemented completely in the app module. This logic has gained some complexity but is not CI tested.

This PR moves part of the logic to QgsVectorLayerEditUtils and add some tests for it. This will also make it accessible for python plugins.

### CODE REVIEW

File: python/core/auto_generated/vector/qgsvectorlayer.sip.in [hunk 1]
```diff
@@ -2636,6 +2636,19 @@ Configuration and logic to apply automatically on any edit happening on this lay
 Returns the manager of the stored expressions for this layer.
 
 .. versionadded:: 3.10
+%End
+
+    bool mergeSelectedFeatures( const QgsAttributes &mergeAttributes, QgsGeometry unionGeometry, QString &errorMessage );
```
Comment (nyalldawson) [inline python/core/auto_generated/vector/qgsvectorlayer.sip.in:2641]: This belongs in QgsVectorLayerEditUtils, not QgsVectorLayer

Comment (domi4484) [reply python/core/auto_generated/vector/qgsvectorlayer.sip.in:2641]: Make sense thanks.
Does it make sense to just merge the selected features or would it better to explicitly pass the features to be merged? I think the second one is more flexible but maybe there are some drawbacks I can't think of...
In the second case should the feature be passed as a list of features or a list of features id?

Comment (nyalldawson) [reply python/core/auto_generated/vector/qgsvectorlayer.sip.in:2641]: > explicitly pass the features to be merged

that one. A list of feature ids will be more flexible IMO.

**Comment by domi4484** [comment]
Test failures seems unrelated... could someone restart them please?

File: src/core/vector/qgsvectorlayereditutils.h [hunk 2]
```diff
@@ -268,6 +268,19 @@ class CORE_EXPORT QgsVectorLayerEditUtils
...
+     * \param mergeFeatureIds id list of features to merge
+     * \param mergeAttributes are the resulting attributes in the merged feature
+     * \param unionGeometry is the resulting geometry of the merged feature
+     * \param errorMessage will be set to a descriptive error message if any occurs
+     *
+     * \returns TRUE if the merge was successful, or FALSE if the operation failed.
+     *
+     * \since QGIS 3.30
+     */
+    bool mergeFeatures( const QgsFeatureIds &mergeFeatureIds, const QgsAttributes &mergeAttributes, QgsGeometry unionGeometry, QString &errorMessage SIP_OUT );
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditutils.h:282]:
```suggestion
    bool mergeFeatures( const QgsFeatureIds &mergeFeatureIds, const QgsAttributes &mergeAttributes, const QgsGeometry &unionGeometry, QString &errorMessage SIP_OUT );
```

File: src/core/vector/qgsvectorlayereditutils.cpp [hunk 3]
```diff
@@ -735,6 +735,119 @@ int QgsVectorLayerEditUtils::addTopologicalPoints( const QgsPointXY &p )
...
+
+  QgsAttributeMap newAttributes;
+  QgsFeatureId mergeFeatureId = FID_NULL;
+  for ( int i = 0; i < mergeAttributes.count(); ++i )
+  {
+    QVariant val = mergeAttributes.at( i );
+
+    bool isDefaultValue = mLayer->fields().fieldOrigin( i ) == QgsFields::OriginProvider &&
+                          mLayer->dataProvider() &&
+                          mLayer->dataProvider()->defaultValueClause( mLayer->fields().fieldOriginIndex( i ) ) == val;
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditutils.cpp:756]: @nyalldawson do we need to take care of `QgsUnsetValue` here?

File: src/core/vector/qgsvectorlayereditutils.cpp [hunk 4]
```diff
@@ -735,6 +735,119 @@ int QgsVectorLayerEditUtils::addTopologicalPoints( const QgsPointXY &p )
   return addTopologicalPoints( QgsPoint( p ) );
 }
 
+bool QgsVectorLayerEditUtils::mergeFeatures( const QgsFeatureIds &mergeFeatureIds, const QgsAttributes &mergeAttributes, QgsGeometry unionGeometry, QString &errorMessage )
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditutils.cpp:738]: Could we alterantively do the following:

- have explicitly a "target feature" being passed into this method
- and 1 or more "merge features"
- we *always* merge into and update that target feature

This basically would get rid of the heuristics in here to determine the target feature by id and some codepaths (when to update or insert)

File: src/core/vector/qgsvectorlayereditutils.cpp (hunk 3 above)
Comment (m-kuhn) [reply src/core/vector/qgsvectorlayereditutils.cpp:756]: This question might be obsolete with the comment below (but still interesting nonetheless).

File: src/core/vector/qgsvectorlayereditutils.cpp [hunk 5]
```diff
@@ -735,6 +735,56 @@ int QgsVectorLayerEditUtils::addTopologicalPoints( const QgsPointXY &p )
...
+  QgsFeatureIds::const_iterator feature_it = mergeFeatureIds.constBegin();
+  for ( ; feature_it != mergeFeatureIds.constEnd(); ++feature_it )
+  {
+    if ( *feature_it != targetFeatureId )
+      mLayer->deleteFeature( *feature_it );
+  }
+
+  // Modify merge feature
+  QgsGeometry mergeGeometry = unionGeometry;
+  mLayer->changeGeometry( targetFeatureId, mergeGeometry );
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditutils.cpp:779]:
```suggestion
  mLayer->changeGeometry( targetFeatureId, unionGeometry );
```

or is there a reason for this copy?

**Review by m-kuhn — APPROVED** [review]
Looks good to me, thank you!

File: src/core/vector/qgsvectorlayereditutils.cpp (hunk 5 above)
Comment (domi4484) [reply src/core/vector/qgsvectorlayereditutils.cpp:779]: According to your previous review I should pass unionGeometry as `const`. But `mLayer->changeGeometry` takes a non-const geometry

File: src/core/vector/qgsvectorlayereditutils.h [hunk 6]
```diff
@@ -268,6 +268,20 @@ class CORE_EXPORT QgsVectorLayerEditUtils
      */
     int addTopologicalPoints( const QgsPointSequence &ps );
 
+    /**
+     * Merge features into a single one.
+     * \param targetFeatureId id of the target feature
+     * \param mergeFeatureIds id list of features to merge
```
[the remark this answers is not in the document]
Comment (domi4484) [reply src/core/vector/qgsvectorlayereditutils.h]: Thanks I changed it

File: src/core/vector/qgsvectorlayereditutils.cpp (hunk 5 above)
Comment (m-kuhn) [reply src/core/vector/qgsvectorlayereditutils.cpp:779]: :-)

**Pull request merged**